### ISSUE TITLE
[1st iteration] Move API credentials to the config.yml file

### ISSUE THREAD

**Issue by jesusgn90** [wazuh/wazuh-kibana-app#1465]
Hi team,

We need to improve the way we manage the Wazuh API entries. We should simplify it adding them to the config.yml file. 

**Proposal**

- It would be an array, each element has URL, port, username, password. 
- The user fills that information, something similar to kibana.yml file.
- The app must show clear instructions if it's started with no API entries.

```yaml
wazuh.hosts:
  - production_env:
      url: http://172.16.1.2
      port: 55000
      username: foo
      password: bar
  - staging_env:
      url: http://localhost
      port: 7600
      username: fuzz
      password: fuzz
```

**Other related changes**

The above change implies some other modifications such as:
- The `.wazuh` index is no longer needed 
- Cluster current status is now different, we must reduce and simplify the way we determine if the cluster is enabled or not.
- Settings view, the form, and all related views must be modified.

**Comment by jesusgn90** [comment]
**Note:** this must be the first iteration, and we can't fully remove the legacy logic so for 3.10.0 the app will accept both mechanisms (soft deprecation so far).

**Comment by juankaromo** [comment]
**Update**

A new field has been added to the component to add APIs. This field will indicate the hostname, and if one is not specified, a random e.g type name will be generated: `host-k54p`.

![image](https://user-images.githubusercontent.com/21195730/59509571-b47da700-8eb1-11e9-876e-1f6c96aa8d2b.png)

When the information is sent to the server for indexing in the `.wazuh`, it is now also added to the config.yml. The entry of this configuration file is also edited when we edit the API.

![image](https://user-images.githubusercontent.com/21195730/59509628-e0009180-8eb1-11e9-9d41-19317108ad02.png)

**Comment by juankaromo** [comment]
**Update**

The current API configuration view will disappear, so you will not be allowed to add or delete API entries from the app. To add them you will have to modify the file `config.yml`.

At this point, and since the APIs are now part of the configuration, I have decided to do the following.

**Proposal**

- A module dedicated to API configuration will be created within the configuration view.
- From here you will only be able to check the connection to the hosts and change the default API.
- To add or remove the configuration file will have to be modified.

![image](https://user-images.githubusercontent.com/21195730/59602918-4979dd80-9108-11e9-8aa2-99ebd1b14bb3.png)

**Migration of index**

Currently the contents of the index `.wazuh` are copied into the configuration file, but only the properties `id, url, port, username, password`. 

Should configurations such as `cluster_info` or `extensions` also be migrated?

**API Identifier**

The APIs in the index had a timestamp to identify them. The name assigned to that host entry will now be taken as the identifier.

**Comment by juankaromo** [comment]
**Update**

We have implemented the possibility of adding a host when there is still none added in the configuration, add it when it already exists and delete it. All this to maintain the current functionality of the APIS table.

For this, a regular expression search is carried out to matchear with the last entry of the yml array and append it to the new host or, if it doesn't exist, to create it.

In the case of deletions, the regular expression is also used to delete the configuration file entry.

The regular expression consists of the name of the entry and as many lines are added as the host object has.

The regular expression with an object of 4 fields would look like this

`/\s*-\s*other\s*:\n\s*\S*\s*\S*\n\s*\S*\s*\S*\n\s*\S*\s*\S*\n\s*\S*\s*\S*/gm`

![image](https://user-images.githubusercontent.com/21195730/59679934-798cb380-91d0-11e9-95d1-6538f332aa1a.png)

**Comment by juankaromo** [comment]
**Update**

Today I have finished doing the complete CRUD to leave the API management from the app identical to how it was done on the `.wazuh` index, but now on the configuration file.

Now we have to approach the scenarios where the index was already created, where we will have to make a migration, from the index to the file and then delete the index.

**Comment by juankaromo** [comment]
**Update** 

Today I've been doing some testing on the processes of checking host connections and application startup checks.

On the other hand, I have been doing a cleaning of the methods that managed the old index `.wazuh`. Some of them are the following:

**Initialize.js**
_checkWazuhIndex_  method now instead of checking that it doesn't exist and creating it, checks if it currently exists and deleted it, after having migrated the existing APIs to config.yml.

**wzWrapper**
All calls made to the `wzWrapper` to get APIS have been replaced by methods to read the configuration file.

**elastic-wrapper**
Methods `createWazuhIndex` and `getWazuhAPIEntries` have been eliminated as they no longer make sense.

**getConfiguration Module**
It has been modified to add two new methods, `getWazuhApiEntries` which gets all the hosts stored in the configuration file, and `getWazuhApiById`, to get a specific API.

**Comment by adri9valle** [comment]
**Update**

Currently, we're saving properly the API information in the `config.yml` file.

```
wazuh.hosts:
  - 1562642463569:
      url: http://172.16.1.2
      port: 55000
      username: foo
      password: YmFy
  - 1562642463569:
      url: http://172.16.1.3
      port: 55000
      username: foo
      password: YmFy      
```

The previous configuration block contains the basic information to authenticate against the API but there is missing information which provokes empty fields in the table with the API entries.


![image](https://user-images.githubusercontent.com/7983984/61220874-8c36d180-a717-11e9-832b-413352061d18.png)


The extensions and cluster/manager information is missing and we need to find out the best way to include this.

The `config.yml` file contents a default extension configuration and in case of this configuration is missing or commented there is a `default-ext.js` file which has the extensions.

I think that we should use the default configuration of the `config.yml` file for all the API entries, and if there is some change in the extensions save this in the local storage.

On the other hand, the cluster information we receive the following information about the cluster/manager when we add a new API:

```
{
     manager: 'master',
     node: 'node01',
     cluster: 'wazuh',
     status: 'enabled' 
 }

```

I think that we should save in the API entry the manager, node and cluster information but not the status, the status should be saved in the local storage or the browser cache and when the status will be checked if changes save it in the browser. An API entry will be:

```
wazuh.hosts:
  - 1562642463569:
      url: http://172.16.1.2
      port: 55000
      username: foo
      password: YmFy
      manager: 'master',
      node: 'node01'
      cluster: 'wazuh'

```

**Comment by adri9valle** [comment]
**Update**

I've fixed an error when trying to check if was https connection with an unexisting function `contains` I replaced it by `includes`. Also, I'm working to save the cluster information.

**Comment by adri9valle** [comment]
**Update**

I've been working on this issue, the new capabilities are now:

- Fixed when accessing the API credentials of the stored APIs in the `config.yml` file. 
- Fixed when checking the connection of an API.
- Checks the connection with the API before selecting it in order to prevent failures.
- Created simple functions in order to prevent duplicated code in several functions.
- Fixed when checking stored API in the cookies.
- Fixed when making API requests.
- Checks the API connection on the React component rendering in order to fetch the cluster or manager information.
- Fixed the fill of the star when selecting a new default API.

There are several fixed points because of the change between the read from the Elasticsearch indices and the `config.yml file`  changes the way to getting the API credentials and a lot of services have been affected.

**Comment by adri9valle** [comment]
**Update**

I've removed several console.log. Also, I fixed the shown error that was appearing when there is not any API entry.

**Comment by adri9valle** [comment]
**Update**

Today I've been adapting and fixing several problems given by the migration from the elasticsearch index to the `config.yml` file.

##### Merged 3.10 and resolved the conflicts.
##### Removed when getting unnecessary the index patterns in the settings controllers provoking an error.
##### Fixed when checking APIs quietly to prevent toast advising of this.
##### Adapted `settings-wizard.js` 
When checking the current API there are some missing fields I have faked this field to prevent temporarily the fails and can continue working in this. Also, I adapted the way to get the APIs entries.
##### Changed the way to check the forms fields,  the function which checks this received the name of the form and with the name it accessed to the class variable instead of pass the object. Now the object is passed directly.
##### Used props instead of state in api-table component in order to fix a binding problem between angularjs and reactjs.
##### When saving a new API entry is returned the generated id in order to set in the API list on the frontend.
#####

**Comment by adri9valle** [comment]
**Update 25.7.19**

Now it's not more used the `base64.encode()` in order to encrypt the password, that's why a user could edit the `config.yml` file by hand and the password will be written as plain text, then the password must not be encoded.  https://github.com/wazuh/wazuh-kibana-app/commit/b53b647100674f53febd92f16c5d0d0c00571eec

**Comment by adri9valle** [comment]
**Update 26.7.19**

The check manager function worked properly until a new API was added from front-end. The problem was that when a new API was added the config.yml file was re-build and the function which composes the APIs entries was setting username instead of user key for the username value, this was provoking that when trying to access to the key user a undefined value was returned.https://github.com/wazuh/wazuh-kibana-app/commit/dc72e68917f801a2f66b5c197129d448da2677ea

**Comment by adri9valle** [comment]
**Update 29.07.19**

In the `initialize.js` when trying to migrate the APIs stored in the `.wazuh` index when there is only one API or anything this works fine, but when there are several APIs the migration process only migrates the first entry and throws an error due to for each API entry were launched several asynchronous functions parallelly and each function tries to write at the same time in the `config.yml` provoking the error `Error: Another process is updating the configuration file` because the file is busy. 

Code that fails:
```
apiEntries.forEach(async x => {
            const host = x._source;
            const added = await updateConfigurationFile.addHost({
              url: host.url,
              port: host.api_port,
              user: host.api_user,
              password: decryptApiPassword(host.api_password)
            });
            if (!added) {
              throw new Error('Error adding Api host to config.yml.');
            }
          });
```

To solve this I split this procedure into several functions and the last function is recursive and checks if there are still APIs to save and if the previous API was saved properly in order to prevent this fail:
https://github.com/wazuh/wazuh-kibana-app/blob/f86b81870b49ba0e0ea6c5470e604ca8ea74e9f4/server/lib/update-configuration.js#L165-L177
Solved: https://github.com/wazuh/wazuh-kibana-app/commit/f86b81870b49ba0e0ea6c5470e604ca8ea74e9f4

**Comment by adri9valle** [comment]
**Update 30.07.2019**

In development mode, the Kibana server entry in a restarting loop because when the APP is initializing writes in the `wazuh-version.json` in the `pluging/wazuh/` path, this provokes that the watcher detects a change and it restarts the Kibana service and again and again... To prevent this the `wazuh-version.json` has been moved to the `optimized/`, also, it has been renamed by `wazuh-registry.json` because `wazuh-version.json` was not very clear. https://github.com/wazuh/wazuh-kibana-app/commit/c24cef7cdc25696656a9adcd8e56644139fc2b64

**Comment by adri9valle** [comment]
**Update 31.07.19**

I changed the ignored file `wazuh-version.json` by `wazuh-registry.json` https://github.com/wazuh/wazuh-kibana-app/commit/2173008f28fc0d96a2e29815a6a90da91653fd38

Now the cluster information is saved and updated as it had done in the `.wazuh` index but now is in the `wazuh-registry.json` https://github.com/wazuh/wazuh-kibana-app/commit/a500cdc820c29d0de62a83001a18e1e18fead425

I've added several debug and error logs messages in the `UpdateRegistry` and `UpdateConfig` services https://github.com/wazuh/wazuh-kibana-app/commit/5da40ad138a89e0c308ae93b4ad3f73e0858f10b

**TODO**

- Now when adding a new API the `checkAPI()` throws an error which says that the API could not be checked.
- Change the way how the cluster information is fetched, now is fetched making an API request but this is no longer necessary because this information is stored and it keeps updated in the `wazuh-registry.json` we need to read the information from there.

**Comment by adri9valle** [comment]
**Update 01.08.19**

The cluster information is now added to the `wazuh-registry.json` when a new API is added or edited from the APP, also, the cluster information is bind from the registry instead of checking the API connection. https://github.com/wazuh/wazuh-kibana-app/pull/1645/commits/c525de67a326eb2ac41f0038657cfdd44800177a

**Comment by adri9valle** [comment]
**Update 02.08.19**

Now while navigating through the app the extensions stored in the cookies and the hosts in the `config.yml` are checked in order to check that there are not any cookie without its corresponding API entry in the `config.yml` preventing the memory flood. https://github.com/wazuh/wazuh-kibana-app/pull/1645/commits/8d4ef470a66ea746507279dd974389c29522a3e2

**TODO**:

- Checks if any API was deleted from the `config.yml` by hand and remove its registry entry in the `wazuh-registry.json`.

**Comment by adri9valle** [comment]
**Update 19.08.19**

I've merged the `3.10-7.2` branch and I've resolved the conflicts, now there is a 3013 error, I need to debug and fix it.
